src/contrib/persistence/Akka.Persistence.Sql.Common/Journal/BatchingSqlJournal.cs

```diff
@@ -1388,17 +1404,59 @@ protected void AddParameter(TCommand command, string paramName, DbType dbType, o
         /// <param name="param">Parameter to customize</param>
         protected virtual void PreAddParameterToCommand(TCommand command, DbParameter param) { }
         
+        /// <summary>
+        /// Scan the buffer and create a chunk of read or write requests based on the type of the first
+        /// request on the queue.
+        ///
+        /// FIX NOTE:
+        /// It is a lot more efficient to sort the requests in the BatchRequest method and
+        /// separate them into two different queues, but we're forced to do this awkward
+        /// conversion because Buffer access level was set as protected and there is
+        /// no guarantee that no concrete implementation is using them.
+        /// -- Greg
+        /// </summary>
         private RequestChunk DequeueChunk(int chunkId)
         {
-            var operationsCount = Math.Min(Buffer.Count, Setup.MaxBatchSize);
-            var array = new IJournalRequest[operationsCount];
-            for (int i = 0; i < operationsCount; i++)
+            // Need a lock here to ensure that buffer doesn't change during this operation
+            lock (_lock)
```

Comment:
This is the very janky code to split the queued requests into either a read request chunk or write request chunk.
Explanation on why it is coded in this peculiar way is written in the method comment.